crates/wasmtime/src/store/future.rs

```diff
@@ -0,0 +1,540 @@
+use crate::store::{StoreInner, StoreOpaque};
+use crate::{AsContext, AsContextMut, StoreContext, StoreContextMut, Trap};
+use std::cell::UnsafeCell;
+use std::future::Future;
+use std::pin::Pin;
+use std::ptr;
+use std::task::{Context, Poll};
+
+type FiberResume = Result<(*mut u8, StoreOpaque<'static>), Trap>;
+type FiberResult = ();
+
+pub struct AsyncState {
+    current_suspend: UnsafeCell<*const wasmtime_fiber::Suspend<FiberResume, (), FiberResult>>,
+    current_poll_cx: UnsafeCell<*mut Context<'static>>,
+}
+
+// Lots of pesky unsafe cells and pointers in this structure. This means we need
+// to declare explicitly that we use this in a threadsafe fashion.
+unsafe impl Send for AsyncState {}
+unsafe impl Sync for AsyncState {}
+
+impl AsyncState {
+    pub fn new() -> AsyncState {
+        AsyncState {
+            current_suspend: UnsafeCell::new(ptr::null()),
+            current_poll_cx: UnsafeCell::new(ptr::null_mut()),
+        }
+    }
+}
```

Comment:
I know this is old PR but just want to ask that did you used `UnsafeCell + ptr` because of self-reference?